src/cpu/flexcpu/simple_dataflow_cpu.hh

```diff
@@ -0,0 +1,242 @@
+        SimpleDataflowCPU* cpu;
+      public:
+        DataPort(const std::string& name_, SimpleDataflowCPU* cpu_):
+            MasterPort(name_, cpu_),
+            cpu(cpu_)
+        { }
+
+        bool recvTimingResp(PacketPtr pkt);
+
+        void revReqRetry();
+    };
+
+    class InstPort : public MasterPort
+    {
+      protected:
+        SimpleDataflowCPU* cpu;
+      public:
+        InstPort(const std::string& name_, SimpleDataflowCPU* cpu_):
+            MasterPort(name_, cpu_),
+            cpu(cpu_)
+        { }
+
+        bool recvTimingResp(PacketPtr pkt);
+
+        void revReqRetry();
+    };
+
+    // END Internal class definitions
+
+
+    // BEGIN Internal state variables
+
+    DataPort _dataPort;
+    InstPort _instPort;
+    vector<SDCPUThread> threads;
+
+    // END Internal state variables
+
+
+    // BEGIN Internal functions
+
+    // TODO some form of speculative state flush function?
+
+
+    // END Internal functions
+
+
+  public:
+    /**
+     * Constructor.
+     */
+    SimpleDataflowCPU(SimpleDataflowCPUParams* params);
+
+    /**
+     * Virtual destructor explicitly declared, so that this class can be
+     * subclassed.
+     */
+    virtual ~SimpleDataflowCPU();
+
+    // BEGIN Member functions (sorted alphabetically)
+
+    /**
+     * Called to set up the state to run a particular thread context. This also
+     * kicks off the first simulation
+     */
+    void activateContext(ThreadID tid) override;
+
+    /**
+     * 
+     */
+    void initiateFetch();
+    // TODO translate the PC, send request, complete decoding.
+
+    /**
+     * Interface calls to retrieve a reference to the port used for data or
+     * instructions respectively.
+     */
+    MasterPort& getDataPort() override;
+    MasterPort& getInstPort() override;
+
+    /**
+     * Called after constructor but before running. (Make sure to call
+     * BaseCPU::init() in this function to let it do its job)
+     */
+    void init() override;
+
+    /**
+     * Event-driven means for other classes to request a timing-based execution
+     * of a given instruction within a given context. Upon completion of the
+     * execution of the instruction, this CPU implementation should notify the
+     * requester by calling the callback function and passing any fault state
+     * resulting from execution.
+     * 
+     * @param inst Reference to an instruction to execute.
+     * @param context Reference to a context interface in which to execute the
+     *  instruction
+     * @param callback_func The function reference for purposes of handling the
+     *  result of the execution.
+     * 
+     * @return Whether or not this request has been acknowledged. If this
+     *  function returns false, the requester should not expect callback_func
+     *  to be called in the future. May return false as a result of 
+     */
+    bool requestExecution(StaticInst& inst, ExecContext& context,
+        std::function<void(Fault)> callback_func);
```

Comment:
I'd tab this over so it's in line with "StaticInst", if that's not too long.

PS: I like this idea of a callback function. I hope it works out!